File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if root == None:
            return "#"
        
        return str(root.val) + "," + self.serialize(root.left) + "," + self.serialize(root.right)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        values = data.split(",")
        self.index = 0

        def preorder():
            if values[self.index] == "#":
                self.index += 1
                return None
            
            root = TreeNode(int(values[self.index]))
            self.index += 1

            root.left = preorder()
            root.right = preorder()

            return root
        
        return preorder()
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root) return "null";
        queue<TreeNode*> que;
        que.push(root);
        string ans = "";
        while (!que.empty()) {
            TreeNode* curr = que.front();
            que.pop();
            if (curr == nullptr) {
                ans += "null,";
            }
            else {
                ans += to_string(curr->val) + ",";
                que.push(curr->left);
                que.push(curr->right);
            }
        }
        return ans;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        string item;
        stringstream ss(data);
        vector<string> values;
        while (getline(ss, item, ',')) {
            values.push_back(item);
        }

        if (values[0] == "null") return nullptr;
        TreeNode* root = new TreeNode(stoi(values[0]));
        queue<TreeNode*> que;
        que.push(root);
        int i=1;
        while (!que.empty()) {
            TreeNode* curr = que.front();
            que.pop();

            string left_val = values[i++];
            if (left_val != "null") {
                curr->left = new TreeNode(stoi(left_val));
                que.push(curr->left);
            } 
            string right_val = values[i++];
            if (right_val != "null") {
                curr->right = new TreeNode(stoi(right_val));
                que.push(curr->right);
            }
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));